pulp_2to3_migration/app/plugin/rpm/migrator.py

```diff
@@ -39,26 +58,135 @@ class RpmMigrator(Pulp2to3PluginMigrator):
+
+class RPMDeclarativeContentMigration(DeclarativeContentMigration):
+    """
+    A pipeline that migrates pre-migrated Pulp 2 rpm content into Pulp 3.
+    """
+
+    def pipeline_stages(self):
+        """
+        Build a list of stages.
+
+        This defines the "architecture" of the content migration to Pulp 3.
+
+        Returns:
+            list: List of :class:`~pulpcore.plugin.stages.Stage` instances
+
+        """
+        pipeline = [
+            self.first_stage,
+            QueryExistingArtifacts(),
+            ArtifactSaver(),
+            QueryExistingContents(),
+            ContentSaver(),
+            InterrelateContent(),
+            RelatePulp2to3Content(),
+        ]
+
+        return pipeline
+
+
+class InterrelateContent(Stage):
+    """
+    Stage for relating Content to other Content.
+    """
+
+    async def run(self):
+        """
+        Relate each item in the input queue to objects specified on the DeclarativeContent.
+        """
+        import pydevd
+        pydevd.settrace('localhost', port=12345, stdoutToServer=True, stderrToServer=True)
+        async for batch in self.batches():
+            module_packages_batch = []
+            modular_packages_batch = []
+            with transaction.atomic():
+                for dc in batch:
+                    if type(dc.content) == pulp3_models.Modulemd:
+                        thru, packages = self.relate_package_to_module(dc)
+                        module_packages_batch.extend(thru)
+                        modular_packages_batch.extend(packages)
+
+                pulp3_models.Package.objects.bulk_update(objs=set(modular_packages_batch),
+                                                        fields=['is_modular'],
+                                                        batch_size=1000)
+
+                ModulemdPackages = pulp3_models.Modulemd.packages.through
+                ModulemdPackages.objects.bulk_create(objs=module_packages_batch,
+                                                     ignore_conflicts=True,
+                                                     batch_size=1000)
+
+            for dc in batch:
+                await self.put(dc)
+
+        #async for dc in self.items():
+        #    if type(dc.content) == Modulemd:
+        #        thru = self.relate_package_to_module(dc)
+
+
+        #    await self.put(dc)
+
+    def relate_package_to_module(self, dc):
+        """
+        Relate a Package to a Module.
+
+        Args:
+            dc (pulpcore.plugin.stages.DeclarativeContent): dc for a Module
+        """
+        ModulemdPackages = pulp3_models.Modulemd.packages.through
+        artifacts_list = dc.content.artifacts
+        # find rpm by nevra
+        # We are relying on the order of the processed DC
+        # RPMs should have passed through ContentSaver stage already
+        pq = Q()
+        for artifact in artifacts_list:
+            nevra = utils.nevra(artifact)
+            pq |= Q(
```

Comment:
In theory, there can be multiple packages withe same nevra but different checksums if we are looking at all the migrated packages in pulp3 :/

Reply:
added logic around this